vermouth/tests/test_molecule.py

```diff
@@ -345,3 +362,623 @@ def test_subgraph_edges(edges_between_molecule, edges_between_selections,
+))
+def test_link_parameter_effector_equal(effector_class, format_spec):
+    """
+    Test that equal LinkParameterEffector compare equal.
+    """
+    n_keys = effector_class.n_keys_asked
+    left_keys = ['A{}'.format(idx) for idx in range(n_keys)]
+    right_keys = copy.copy(left_keys)  # Let's be sure the id is different
+    left = effector_class(left_keys, format_spec=format_spec)
+    right = effector_class(right_keys, format_spec=format_spec)
+    assert left == right
+
+
+@pytest.mark.parametrize('effector_class', (
+    vermouth.molecule.ParamDistance,
+    vermouth.molecule.ParamAngle,
+    vermouth.molecule.ParamDihedral,
+    vermouth.molecule.ParamDihedralPhase,
+))
+@pytest.mark.parametrize('format_right, format_left', itertools.combinations(
+    (None, '.2f', '0.3f', ), 2
+))
+def test_link_parameter_effector_diff_format(effector_class, format_left, format_right):
+    """
+    Test that LinkParameterEffector compare different if they have different format.
+    """
+    n_keys = effector_class.n_keys_asked
+    left_keys = ['A{}'.format(idx) for idx in range(n_keys)]
+    right_keys = copy.copy(left_keys)  # Let's be sure the id is different
+    left = effector_class(left_keys, format_spec=format_left)
+    right = effector_class(right_keys, format_spec=format_right)
+    assert left != right
+
+
+@pytest.mark.parametrize('effector_class', (
+    vermouth.molecule.ParamDistance,
+    vermouth.molecule.ParamAngle,
+    vermouth.molecule.ParamDihedral,
+    vermouth.molecule.ParamDihedralPhase,
+))
+def test_link_parameter_effector_diff_keys(effector_class):
+    """
+    Test that LinkParameterEffector compare different if they have different keys.
+    """
+    n_keys = effector_class.n_keys_asked
+    left_keys = ['A{}'.format(idx) for idx in range(n_keys)]
+    right_keys = ['B{}'.format(idx) for idx in range(n_keys)]
+    left = effector_class(left_keys)
+    right = effector_class(right_keys)
+    assert left != right
+
+
+@pytest.mark.parametrize('left_class, right_class', itertools.combinations((
+    vermouth.molecule.ParamDistance,
+    vermouth.molecule.ParamAngle,
+    vermouth.molecule.ParamDihedral,
+    vermouth.molecule.ParamDihedralPhase,
+), 2))
+def test_link_parameter_effector_diff_class(left_class, right_class):
+    """
+    Test that LinkParameterEffector compare different if they have different classes.
+    """
+    left_n_keys = left_class.n_keys_asked
+    left_keys = ['A{}'.format(idx) for idx in range(left_n_keys)]
+    left = left_class(left_keys)
+
+    right_n_keys = right_class.n_keys_asked
+    right_keys = ['B{}'.format(idx) for idx in range(right_n_keys)]
+    right = right_class(right_keys)
+
+    assert left != right
+
+
+@st.composite
+def attribute_dict(draw, min_size=0, max_size=None):
+    keys = st.one_of(st.text(), st.integers(), st.none())
+    bases = (st.text(), st.integers(), st.floats(), st.none())
+    lists = st.lists(st.one_of(*bases, attribute_dict(max_size=1)), max_size=3)
```

Comment:
This reduces to st.Text for failing tests since that's the first element of bases. Maybe change that order so None or int is first?